gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -252,24 +256,50 @@ private WriteObjectResponse doResumableUpload() throws IOException {
 
       boolean objectFinalized = false;
       while (!objectFinalized) {
-        WriteObjectRequest insertRequest;
+        WriteObjectRequest insertRequest = null;
         if (dataChunkMap.size() > 0 && dataChunkMap.lastKey() >= writeOffset) {
           insertRequest = buildRequestFromBufferedDataChunk(dataChunkMap, writeOffset);
           writeOffset += insertRequest.getChecksummedData().getContent().size();
         } else {
-          ByteString data =
-              ByteString.readFrom(
-                  ByteStreams.limit(pipeSource, MAX_BYTES_PER_MESSAGE), MAX_BYTES_PER_MESSAGE);
-          dataChunkMap.put(writeOffset, data);
           if (dataChunkMap.size() >= channelOptions.getNumberOfBufferedRequests()) {
-            dataChunkMap.remove(dataChunkMap.firstKey());
+            /*
+             If dataChunkMap is full, get the committedWriteOffset. This will make a API call to the
+             server and add latency in this path/context. The upload is happening asynchronously in
+             a different context. So this API will not add latency to overall upload. It will only
+             throttle the upstream write call, which is fine.
+            */
+            committedWriteOffset = getCommittedWriteSizeWithRetries(uploadId);
+            logger.atFinest().log(
+                "Fetched committedWriteOffset: size:%d, numBuffers:%d, writeOffset:%d, committedWriteOffset:%d",
+                dataChunkMap.size(),
+                channelOptions.getNumberOfBufferedRequests(),
+                this.writeOffset,
+                committedWriteOffset);
+
+            // check and remove chunks from dataChunkMap
+            while (dataChunkMap.size() > 0 && dataChunkMap.firstKey() < committedWriteOffset) {
+              logger.atFinest().log(
+                  "clearing dataChunkMap one buffer at a time, size: %d, firstKey:%d, committedwriteOffset:%d",
+                  dataChunkMap.size(), dataChunkMap.firstKey(), committedWriteOffset);
+              dataChunkMap.remove(dataChunkMap.firstKey());
```

Comment:
If we have a commit offset, why not delete all data upto that offset, instead of just the first key ?

This is add additional calls to server, and sub-optimal buffer clear performance

Reply:
nvm, misred while loop for if